### PR TITLE
Chore: improve context key injection

### PR DESCRIPTION
**Pull request by Laurin-W**
Hey @Ruben
addressing issue https://github.com/comunica/comunica/issues/949, I created this PR.

I had to add an ignoreComponent record to `.componentsjs-generator-config.json`'s as it does not seem to support Partial correctly which is used here for context type support in the IDE.

I hope this has caught the right direction and happy for feedback :)

### CODE REVIEW

**Comment by Laurin-W** [comment]
And there seems to be an issue with `QuerySparql-solid-test.ts` on node 14 but I can't reproduce it on my local machine..

File: packages/actor-init-query/lib/ActorInitQueryBase.ts [hunk 1]
```diff
@@ -28,9 +29,27 @@ export class ActorInitQueryBase extends ActorInit implements IActorInitQueryBase
...
+   * Create new ActorInitQueryBase object.
+   * @param args.contextKeyShortcutsExtensions Array of `contextKeyShortcuts` that are merged
+   *   with the `contextKeyShortcuts` field. This allows adding shortcuts to the defaults.
+   * @throws When duplicate keys are present in `args.contextKeyShortcuts`
+   *  and `args.contextKeyShortcutsExtensions`.
+   */
   public constructor(args: IActorInitQueryBaseArgs) {
+    // Add additional contextKeyShortcuts.
+    args.contextKeyShortcutsExtensions?.forEach(extensionShortcuts => {
+      // Throw, f there are duplicate keys that are to be added to `contextKeyShortcuts`.
```
Comment (rubensworks) [inline packages/actor-init-query/lib/ActorInitQueryBase.ts:47]:
```suggestion
      // Throw, if there are duplicate keys that are to be added to `contextKeyShortcuts`.
```

File: packages/actor-init-query/lib/ActorInitQueryBase.ts [hunk 2]
```diff
@@ -126,5 +146,25 @@ export interface IActorInitQueryBaseArgs extends IActorInitArgs {
...
    * }}
    */
-  contextKeyShortcuts: Record<string, string>;
+  contextKeyShortcuts: Record<string, string> | Partial<Record<keyof QueryContext, string>>;
+  /**
+   * An array of `contextKeyShortcuts` that are to be appended to the (default) `contextKeyShortcuts`
+   * (which are by default injected by component.js).
+   *
+   * The appending happens in the constructor call. Conflicting keys will cause an error.
+   * If you extend `ActorInitQueryBase` and want to add custom shortcuts, do so as follows:
```
Comment (rubensworks) [inline packages/actor-init-query/lib/ActorInitQueryBase.ts:155]: This is what we need indeed! But I would go a step further, and make it so that the TS code below is not needed anymore.
Using Components.js annotations magic, we should be able to define these extensions via JSDoc annotations, similar to [here](https://github.com/comunica/comunica/blob/master/packages/actor-init-query/lib/ActorInitQueryBase.ts#L97-L128).

**Review by rubensworks — CHANGES_REQUESTED** [review]
Nice work! We're on the right track, but I think we need one more thing for better developer experience.

**Comment by rubensworks** [comment]
Opened an issue for the test failure, since I saw it popping up in another unrelated PR: https://github.com/comunica/comunica/issues/1106

File: packages/actor-init-query/lib/ActorInitQueryBase.ts (hunk 2 above)
Comment (Laurin-W) [reply packages/actor-init-query/lib/ActorInitQueryBase.ts:155]: I'm not quite sure how to achieve this with the annotation magic referenced here (and couldn't find documentation).

Since the goal after all would be to extend the annotated default context (in the current case with `contextKeyShortcutsExtensions`), I am not sure how this would work for chained inherits (without overriding `contextKeyShortcutsExtensions` which would make the `Record<string, string>[]` array useless).

Am I getting that wrong or missing something?

Comment (rubensworks) [reply packages/actor-init-query/lib/ActorInitQueryBase.ts:155]: You're right, this isn't possible yet using Components.js annotations.
We'd have to extend the Components.js generator to make this array append logic possible.
Instead of doing that, it would be better to just extend Components.js to extend JSON objects.

But in any case, that's beyond the scope of this PR, so let's leave that for future work. So you can ignore my comment here :-)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -9,8 +9,6 @@ jobs:
         os: [ubuntu-latest]
         node-version:
           - 14.x
-          - 16.x
```
Comment (rubensworks) [inline .github/workflows/ci.yml:12]: Change not needed?

File: packages/actor-init-query/test/HttpServiceSparqlEndpoint-test.ts [hunk 4]
```diff
@@ -48,6 +49,16 @@ const argsDefault = {
 };
 
 describe('HttpServiceSparqlEndpoint', () => {
+  let originalCluster: typeof cluster;
```
Comment (rubensworks) [inline packages/actor-init-query/test/HttpServiceSparqlEndpoint-test.ts:52]: Is this a fix for #1106? If so, could we move this to a separate PR?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (Laurin-W) [reply .github/workflows/ci.yml:12]: Hey Ruben, that was just for testing purposes since I wasn't able to reproduce the issues locally (and didn't want the succeeding tests to be re-run to save some resources).

Comment (rubensworks) [reply .github/workflows/ci.yml:12]: Ah ok. Just ping me when this is ready for review :)

File: packages/actor-init-query/test/HttpServiceSparqlEndpoint-test.ts (hunk 4 above)
Comment (Laurin-W) [reply packages/actor-init-query/test/HttpServiceSparqlEndpoint-test.ts:52]: Yep, sorry! Was experimenting with that last night since I needed a test execution environment.

File: packages/actor-init-query/lib/ActorInitQueryBase.ts (hunk 1 above)
Comment (Laurin-W) [reply packages/actor-init-query/lib/ActorInitQueryBase.ts]: addressed with https://github.com/comunica/comunica/pull/1102/commits/4a7853fb9bd44530c5bcabb271a85f8086cc7f89

File: packages/actor-init-query/test/HttpServiceSparqlEndpoint-test.ts (hunk 4 above)
Comment (Laurin-W) [reply packages/actor-init-query/test/HttpServiceSparqlEndpoint-test.ts:52]: Will push reversion once the fix (https://github.com/comunica/comunica/pull/1109) is merged.
Or would you recommend to push it immediately?

Comment (rubensworks) [reply packages/actor-init-query/test/HttpServiceSparqlEndpoint-test.ts:52]: It's ok, we can wait :)

**Comment by Laurin-W** [comment]
Hey @rubensworks! This should be good to go then :)

**Comment by rubensworks** [comment]
Good stuff, thanks @Laurin-W!
